consensus/istanbul/proxy/proxy_handler.go

```diff
@@ -0,0 +1,352 @@
+// Copyright 2017 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package proxy
+
+import (
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus"
+	"github.com/ethereum/go-ethereum/consensus/istanbul"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/enode"
+)
+
+// This struct defines the handler that will manage all of the proxies and
+// validator assignments to them
+type proxyHandler struct {
+	lock    sync.RWMutex // protects "running" and "p2pserver"
+	running bool         // indicates if `run` is currently being run in a goroutine
+
+	loopWG sync.WaitGroup
+	quit   chan struct{}
+
+	addProxies    chan []*istanbul.ProxyConfig // This channel is for adding new proxies specified via command line or rpc api
```

Comment:
As mentioned in another note, `doc comments` go on the line before, not at the end.

(or maybe that works too, but not the common idiom). See https://golang.org/doc/effective_go.html#commentary

Reply:
Done.